include/votca/xtp/davidsonsolver.h

```diff
@@ -0,0 +1,296 @@
+/*
+ * Copyright 2009-2019 The VOTCA Development Team (http://www.votca.org)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef __VOTCA_TOOLS_DAVIDSON_SOLVER_H
+#define __VOTCA_TOOLS_DAVIDSON_SOLVER_H
+
+#include <iostream>
+#include <stdexcept>
+#include <chrono>
+
+#include <votca/xtp/eigen.h>
+
+#include <votca/xtp/orbitals.h>
+#include <votca/xtp/qmstate.h>
+#include <votca/xtp/rpa.h>
+#include <votca/xtp/threecenter.h>
+
+using boost::format;
+using std::flush;
+
+namespace votca {
+namespace xtp {
+
+/**
+* \brief Use Davidson algorithm to solve A*V=E*V
+
+**/
+
+class DavidsonSolver {
+
+ public:
+  DavidsonSolver(ctp::Logger &log);
+
+  void set_iter_max(int N) { this->iter_max = N; }
+  void set_tolerance(double eps) { this->tol = eps; }
+  void set_max_search_space(int N) { this->max_search_space = N; }
+
+  void set_correction(std::string method);
+
+  Eigen::VectorXd eigenvalues() const { return this->_eigenvalues; }
+  Eigen::MatrixXd eigenvectors() const { return this->_eigenvectors; }
+
+  template <typename MatrixReplacement>
+  void solve(MatrixReplacement &A, int neigen, int size_initial_guess = 0) {
+
+    std::chrono::time_point<std::chrono::system_clock> start, end;
+    start = std::chrono::system_clock::now();
+
+    CTP_LOG(ctp::logDEBUG, _log) << ctp::TimeStamp() << " Davidson Solver"
+                                 << flush;
+
+    switch (this->davidson_correction) {
+
+          case CORR::DPR:
+            CTP_LOG(ctp::logDEBUG, _log) << ctp::TimeStamp() 
+            << " DPR Correction" << flush;
+            break;
+
```

Comment:
default case handling for other inputs, are the options checked before?

Reply:
If I understood the input parsing correctly, the default are set in `bse.h` . The input file is parsed in `gwbse.cc` with `ìfExistsReturnElseReturnDefault`. The correction term is (DPR or Olsen) is checked with  `ifExistAndInListReturnElseThrowRuntimError`. Is that ok ?

Reply:
yes